src/main/scala/edu/berkeley/ce/sparkrocks/RockSlicer.scala

```diff
@@ -121,16 +124,16 @@ object RockSlicer {
       // Convert the list of rock blocks to JSON with vertices, normals and connectivity in
       // format easily converted to vtk by rockProcessor module
       val vtkBlocks = squeakyClean.map(BlockVTK(_))
-      val jsonVtkBlocks = vtkBlocks.map(JsonToVtk.blockVtkToMinimalJson)
-      jsonVtkBlocks.saveAsTextFile(arguments.vtkOut)
+      val jsonVtkBlocks = JsonToVtk.blockVtkSeqToMinimalJson(vtkBlocks.collect())
+      Files.write(Paths.get(arguments.demOut), jsonVtkBlocks.getBytes(StandardCharsets.UTF_8))
     }
 
     if (arguments.demOut != "") {
       // Convert the list of rock blocks to JSON format appropriate for DEM analysis and
       // save this to a file
       val demBlocks = squeakyClean.map(BlockDem(_))
-      val jsonDemBlocks = demBlocks.map(JsonToDem.blockDemToMinimalJson)
-      jsonDemBlocks.saveAsTextFile(arguments.demOut)
+      val jsonDemBlocks = JsonToDem.blockDemSeqToMinimalJson(demBlocks.collect())
```

Comment:
So if I'm understanding things correctly: Before we used `map` to convert each block to JSON, then just had spark write out a text file for us.

Now we want to save a single file on the driver node so we're using `collect` for that. Seems like that could work, but it looks like here we're doing all of the JSON conversion on the driver node. Wouldn't it be better to keep a call to `map` to convert each block to JSON on the full cluster, then `collect` the RDD of strings and write the file?

The other issue is that this makes SparkRocks less "standard" when it comes to using Spark. The `saveAsTextFile` method they provide assumes you're running on something like HDFS where you don't want to have a single machine deal with all of the data.

But I guess the flipside is that if downstream tools are not on Spark, which is almost definitely going to be the case, this change makes things much easier, so maybe you're right that we should be less Spark focused and more focused on interfacing with other practical geo tools.

Reply:
Yeah, I was thinking about this too. Seems like it would also make writing the output a bottleneck while making it hard to see which blocks were generated by which process. I've been looking, but I can't seem to find a way to convert the partitions into JSON strings before calling `saveAsTextFile`. 

Would prefer to keep it along inline with "standard" Spark use, but not sure how we could have the executors write local partitions as JSON sequences instead of a JSON objects per line. Do you have any ideas about that?

Reply:
So the main issue is that, for each block, we want to know which of the load-balancing regions within the rock mass it originally came from? If we want to keep our code "Spark-ish" and avoid the `collect` there are some things we might try.

I guess the nuclear option would be to have each block explicitly track some kind of region ID and include that in its JSON representation. The downside is this would require lots of small changes to the `Block` code.

Assuming each partition of the Spark RDD corresponds to one of our artificial load balancing regions, there are some other possibilities. This is still true right?

We could try using `mapPartitions` rather than `map` to convert each RDD partition to a sequence of strings, so we'd go from an RDD of blocks to an RDD of string sequences, and then hopefully Spark writes to the file at the granularity of these sequences rather than individual blocks.

Maybe we'd have to use `mapPartitionsWithIndex` and/or `groupByKey` if we also need a more explicit indication of which partition a block belongs to.

Thoughts on all this?

Reply:
We don't necessarily need to know which region the rock mass originally came from, though that would be nice. I'd say no to the nuclear option; this is overkill and wouldn't really add to what's already there.

Yes, each RDD still corresponds to one of the artificial load balancing zones. This worked quite well for writing the output last time as `saveAsTextFile` automatically output a text file for each RDD partition. The issue that I keep running into now is that there isn't a natural way to have each partition of a RDD transformed into a JSON array and then writing that to file.

I tried using `mapPartitions` to convert each partition to a JSON array and then use `saveAsTextFile`, but this ends up giving you a text file that just has each character on a newline. From what I can tell, the way Spark normally deals with JSON is to assume that each line will contain a JSON object--this is waht we were doing before.

With that, if we want to stay along the lines of Spark, we'd have to post process the outputs and create a proper JSON array by going through the generated output files line by line. This would be fine; it just adds a post-processing step that seems a bit of a pain. 

Not sure what the best path forward is. I liked being able to feed the outputs directly into the analysis software I'm using, but that might be a shortsighted solution...

Reply:
Ok. Maybe I understand now, still not 100% sure.

I think the fundamental issue is that you want a situation where you can deal pretty directly with an RDD's internal partitions, which doesn't align with Spark's intended abstraction. We have an instance of `RDD[Block]` and Spark wants us to treat that as a single collection, despite the fact that it is likely partitioned among nodes in a cluster.

The way I see it, there are three options.

**Option 1:** Keep things the way they were before. SparkRocks produces a collection of blocks, and we can express this in a JSON file where each block is on its own line.

In my opinion, this may actually be the best option. It is the most idiomatic in regards to Spark practices and probably also has the best performance, since we can let each node in the cluster construct JSON representations in parallel.

I realize I'm out of touch with what you're trying to do downstream, but I'm also not sure I agree with the idea of exporting a representation where the blocks are explicitly separated into groups. First, the fact that we have separated the rock mass into synthetic regions for load balancing is an aspect of our internal implementation. It doesn't necessarily need to be reflected in the interface we expose to users. And if for some reason we wanted to later change the SparkRocks load balancing implementation, we're free to do this without any disruption to what users may expect as output. To me, it seems similar to the motivation for keeping certain fields and methods `private` in a class.

Also, from my perspective it doesn't seem particularly onerous to expect downstream tools to do their own post processing if they want blocks grouped into regions. If we choose to separate the blocks within SparkRocks, who is to say that our choice of grouping logic is the right one? Maybe one user is okay with our strategy but another wants to separate blocks using planes along the z-axis? I liked the idea that SparkRocks doesn't express an opinion here. If downstream users want partitioning, they can do the post processing that exactly matches their needs, and they don't need to work around SparkRocks if it were to impose its own grouping scheme that doesn't match the user's needs.

That said, you're the one who'll actually be using this, and you obviously know a lot more about what kinds of things people would do downstream from SparkRocks, so I'll defer to you if you think we should output a JSON representation where blocks are grouped.

**Option 2:** Merge in the changes you've made here. This code won't perform as well as Option 1, but it will do what you want. And even if it isn't quite what a typical Spark application would do, we're at least using common functions that any Spark person would understand.

**Option 3:** There's actually a way to get each partition of the RDD as its own "line" of JSON in the text file Spark produces, using the [glom](https://spark.apache.org/docs/latest/api/scala/index.html#org.apache.spark.rdd.RDD@glom():org.apache.spark.rdd.RDD[Array[T]]) method. You can write something like this:
```scala
val jsonBlocks = demBlocks.glom.map(Json.blockArrayToJson).saveAsTextFile(...)
```

`glom` gets you from something like `RDD[Block]` to `RDD[Array[Block]]` where each array corresponds to one of the original partitions. Note that `Array` is not an instance of `Seq`, so we'd probably have to replace the function for converting a `Seq[Block]` to JSON to one that takes `Array[Block]`.

I tested out a simple version of this on a local Spark install, it should work.

This will have better performance than Option 2 since there's no call to `collect`. But I get the feeling `glom` is a little off the beaten path, given the fact that it's not discussed in Spark's RDD Programming guide and its documentation is a bit sparse. But I'd probably prefer Option 3 to Option 2.

Thoughts?

Reply:
Sorry for the slow reply! You make a really good argument for Option 1. I agree that Option 2 is a no-go. 

I like `glom` since it automatically does that for each RDD, but then it becomes a bit more convoluted in terms of how you might post-process each individual JSON array. So, perhaps it makes more sense to write a post-processing script in Python to take all the outputs and then turn them into a single JSON array. 

With that then, I guess the only thing we should keep in this pull request are the updates to the sbt and Travis configs?

Reply:
Yep, if you're ok with Option 1 then I think all we need is the sbt and Travis stuff.

Reply:
Okay, that was a majorly slow reply! Just changed it back to the more Spark fundamentalist approach with `saveAsTextFile`, so should be ready to merge with changes only for sbt and Travis config.

Reply:
No worries, I think it looks good now.